msrestazure/azure_active_directory.py

```diff
@@ -525,3 +527,52 @@ def set_token(self, response_url):
             raise_with_traceback(AuthenticationError, "", err)
         else:
             self.token = token
+
+
+# Constants related to AAD-based authentication methods.
+_TOKEN_ENTRY_TOKEN_TYPE = 'tokenType'
+_ACCESS_TOKEN = 'accessToken'
+
+
+class AdalAuthentication(msrest.authentication.Authentication):
+
+    """Base class for adal-derived authentication."""
+
+    def __init__(self, client_id="04b07795-8ddb-461a-bbee-02f9e1bf7b46",
+                 tenant="common",
+                 auth_endpoint="https://login.microsoftonline.com",
+                 resource="https://management.core.windows.net/"):
+        """Handle details common to adal."""
+        super(AdalAuthentication, self).__init__()
+        self.client_id = client_id  # Default value is xplat client ID.
+        self.authority = urljoin(auth_endpoint, tenant)
+        self.resource = resource
+
+    # @abc.abstractmethod if Python 2 wasn't supported.
+    def acquire_token(self, context):
+        """Override with code that returns an adal.acquire_*() call."""
+        raise NotImplementedError
+
+    def signed_session(self):
+        """Return a signed session."""
+        session = super(AdalAuthentication, self).signed_session()
+        context = adal.AuthenticationContext(self.authority)
+        token_entry = self.acquire_token(context)
+        header = " ".join([token_entry[_TOKEN_ENTRY_TOKEN_TYPE],
+                                token_entry[_ACCESS_TOKEN]])
+        session.headers['Authorization'] = header
+        return session
+
+
+class AdalUserPassCredentials(AdalAuthentication):
+
+    """Authenticate with AAD using a username and password."""
+
+    def __init__(self, username, password, **kwargs):
+        super(AdalUserPassCredentials, self).__init__(**kwargs)
+        self.username = username
+        self.password = password
```

Comment:
This implies that credentials are not tested at the instance creation, but at the instance usage. This is different from current `UserPassCredentials` behavior.
I think we might want to keep the current behavior, but I'm open to discussion (@annatisch)